design/proposal-controller-code-generation.md

```diff
@@ -0,0 +1,129 @@
+
+### AWS Service Operator (ASO)
+The [pipeline architecture][aso-v2-architecture] of the next iteration of [Amazon Service Operator][amazon-aso] appears to be very similar to what we are considering, leapfrogging kubebuilder and working directly with `controller-tools` to generate controller code. The existing ASO uses CloudFormation for Provider API metadata discovery; it’s unclear from the architecture doc if they plan to continue to use CloudFormation or another API description: “The first code generation phase consumes model information from a canonical source of truth about an AWS service”.
+
+### Magic Modules
+[Magic Modules] is a Google project, written in Ruby, that does low-level code generation, translating Google’s discovery format to bindings for Terraform and Ansible.
+
+## Kubeform
+[KubeForm] generates CRDs using Hashicorp’s [terraform-config-inspect] library to extract types from Terraform providers, rendering them to CRDs and boilerplate Go types via Go templates. Under the hood the controller (KFC) uses the generated types to for k8s API interactions and templates out Terraform manifests for resources requiring reconciliation, [directly exec()’ing `terraform apply`][kubeform-terraform-apply]. This project does not appear to be actively maintained.
+
+### Pulumi
+[Pulumi] is internally implemented as a wrapper on top of Terraform providers; it is effectively a Terraform SDK. Pulumi not only uses Terraform to map out provider types and methods, but also links in Terraform types and interacts with Terraform's state file data structure through an adapter layer. A lot of the machinery in pulumi exists to present a consistent interface to Terraform across other supported languages inc javascript, python and .NET.
+
+[resource.Managed interface]: https://godoc.org/github.com/crossplane/crossplane-runtime/pkg/resource#Managed
+[OpenAPI Directory]: https://github.com/APIs-guru/openapi-directory
+[Alibaba API Explorer]: https://api.aliyun.com/#/
+[directly supports OpenAPI]: https://github.com/Azure/azure-rest-api-specs
+[terraform-alicloud]: https://github.com/terraform-providers/terraform-provider-alicloud
+[terraform-aws]: https://github.com/terraform-providers/terraform-provider-aws
+[terraform-azure]: https://github.com/terraform-providers/terraform-provider-azurerm
+[terraform-gcp]: https://github.com/terraform-providers/terraform-provider-google
+[terraform-config-inspect]: https://github.com/hashicorp/terraform-config-inspect/
+[aws-cloudformation]: https://aws.amazon.com/cloudformation/
+[gcp-config-connector]: https://cloud.google.com/config-connector/docs/overview
+[azure-resource-manager]: https://docs.microsoft.com/en-us/azure/azure-resource-manager/management/overview
+[alibaba-ros]: https://www.alibabacloud.com/product/ros
```

Comment:
For ROS, we normally use https://github.com/oam-dev/cloud-provider/tree/master/alibabacloud/ros because it's declarative.